compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```diff
@@ -425,71 +456,77 @@ MMAScheduleAttr::getContractionLayout(vector::ContractionOp contractOp) const {
-  SmallVector<int64_t, 2> cBatchOrder = {0, 1};
-  SmallVector<int64_t, 2> cThreadBasis = cCounts.thread;
-
-  auto cLayout = NestedLayoutAttr::get(
-      getContext(), cSubgroupPerWorkgroup, cSubgroupOrder, cBatchesPerSubgroup,
-      cBatchOrder, cCounts.outer, cOrders.outer, cCounts.thread, cOrders.thread,
-      cCounts.element, cOrders.element, subgroupBasis, cThreadBasis);
+  // Note that no matter how we permute/transpose the input contraction problem,
+  // the way we view the hardware warps/threads remain the same--that is, from
+  // hardware's perspective, a single thread has the same warp and thread ID no
+  // matter what part of the contraction data it works on. Similarly here, we
+  // are delinearizing the linearized GPU hardware thread ID into a n-D
+  // concatenated logical warp+thread using the basis, and they should remain
+  // the same for all A/B/C matrix.
+  SmallVector<int64_t, 2> subgroupBasis = {getSubgroupMCount(),
+                                           getSubgroupNCount()};
+  SmallVector<int64_t, 2> threadBasis = cCounts.thread;
+
+  auto cLayout = permuteAndCreateNestedLayout(
+      context, cPermute,
+      /*subgroupCount=*/{getSubgroupMCount(), getSubgroupNCount()},
+      /*subgroupOrder=*/{0, 1},
+      /*batchCount=*/{getSubgroupMTileCount(), getSubgroupNTileCount()},
+      /*batchOrder=*/{0, 1}, /*outerCount=*/cCounts.outer,
+      /*outerOrder=*/cOrders.outer, /*threadCount=*/cCounts.thread,
+      /*threadOrder=*/cOrders.thread,
+      /*elementCount=*/cCounts.element, /*elementOrder=*/cOrders.element,
+      subgroupBasis, threadBasis);
 
   // A matrix layout
   MFMAAttr::SingleSubgroupLayout aCounts =
       mfmaAttr.getASingleSubgroupLayoutCount();
   MFMAAttr::SingleSubgroupLayout aOrders =
       mfmaAttr.getASingleSubgroupLayoutOrder();
 
-  SmallVector<int64_t, 2> aSubgroupPerWorkgroup = {getSubgroupMCount(), 1};
-  SmallVector<int64_t, 2> aBatchesPerSubgroup = {getSubgroupMTileCount(),
-                                                 getSubgroupKTileCount()};
-  SmallVector<int64_t, 2> aSubgroupOrder = {0, 1};
-  SmallVector<int64_t, 2> aBatchOrder = {0, 1};
-  SmallVector<int64_t, 2> aThreadBasis = aCounts.thread;
-
-  auto aLayout = NestedLayoutAttr::get(
-      getContext(), aSubgroupPerWorkgroup, aSubgroupOrder, aBatchesPerSubgroup,
-      aBatchOrder, aCounts.outer, aOrders.outer, aCounts.thread, aOrders.thread,
-      aCounts.element, aOrders.element, subgroupBasis, aThreadBasis);
+  auto aLayout = permuteAndCreateNestedLayout(
+      context, aPermute,
+      /*subgroupCount=*/{getSubgroupMCount(), 1},
+      /*subgroupOrder=*/{0, 1},
+      /*batchCount=*/{getSubgroupMTileCount(), getSubgroupKTileCount()},
+      /*batchOrder=*/{0, 1}, /*outerCount=*/aCounts.outer,
+      /*outerOrder=*/aOrders.outer, /*threadCount=*/aCounts.thread,
+      /*threadOrder=*/aOrders.thread,
+      /*elementCount=*/aCounts.element, /*elementOrder=*/aOrders.element,
+      subgroupBasis, threadBasis);
```

Comment:
I'm not 100% sure the thread basis is the same for all fragments (A/B/C) of the MMA. The implicit shuffles of the MMA intrinsics makes it so that there doesn't have to be any direct relationship between the thread basis of `A` and `B` AFAIU. It looks like this works for the cases we currently support, but I think the correct solution here is to transpose/permute the `aCounts.thread` with `aOrders.thread` the same way I suggested above for C.